ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py

```diff
@@ -3058,3 +3058,76 @@ def test_tensorflow_Elu(
         features=x[0],
         name=name,
     )
+
+
+@st.composite
+def explicit_paddings(draw):
+    x, filters, strides, dilations, data_format = _x_and_filters(draw)
+
+    padding = draw(
+        st.lists(
+            st.tuples(
+                st.integers(min_value=0, max_value=3),
+                st.integers(min_value=0, max_value=3),
+            ),
+            min_size=2,
+            max_size=2,
+        )
+    )
+
+    if data_format == 'NCHW':
+        padding.append((0, 0))
+    else:
+        padding.insert((0, 0), (0, 0))
+
+    return padding
+
+
+@handle_frontend_test(
+    fn_tree="tensorflow.raw_ops.Conv2D",
+    x_f_d_df=_x_and_filters(
+        dtypes=helpers.get_dtypes("float", full=False),
+        data_format=st.sampled_from(["NHWC", "NCHW"]),
+        padding=st.sampled_from(["SAME", "VALID", "EXPLICIT"]),
+        explicit_paddings=explicit_paddings(),
```

Comment:
I meant call `x_f_d_df=draw(_x_and_filters(...))` inside a composite strategy, take the `data_format` from `x_f_d_df` and then generate `explicit_paddings`.

Reply:
@AnnaTz  you want me to create a different @st.composite function right, where i will use the explicit_padding
or anything else.
this is the current function that i made in the raw_ops file 

```
@st.composite
def explicit_paddings(draw):
    x, filters, strides, dilations, data_format = _x_and_filters(draw)

    padding = draw(
        st.lists(
            st.tuples(
                st.integers(min_value=0, max_value=3),
                st.integers(min_value=0, max_value=3),
            ),
            min_size=2,
            max_size=2,
        )
    )

    if data_format == "NHWC":
        padding = [(0, 0), (padding[0][0], padding[0][1]), (padding[1][0], padding[1][1]), (0, 0)]
    else:
        padding = [(0, 0), (0, 0), (padding[0][0], padding[0][1]), (padding[1][0], padding[1][1])]

    if data_format == "NCHW":
        padding.insert(0, (0, 0))
        padding.insert(1, (0, 0))

    return padding
```

Reply:
@AnnaTz  are you suggesting something like this

```@st.composite
def conv2d_inputs(draw):
x_f_d_df=draw(_x_and_filters(
        dtypes=helpers.get_dtypes("float", full=False),
        data_format=st.sampled_from(["NHWC", "NCHW"]),
        padding=st.sampled_from(["SAME", "VALID", "EXPLICIT"]),
        explicit_paddings=explicit_paddings(),
        type="2d",
        dilation_min=1,
        dilation_max=1,
        ))
  explicit_paddings = None
    padding = draw(st.sampled_from(["SAME", "VALID", "EXPLICIT"]))
    if padding == "EXPLICIT":
        padding = draw(st.lists(st.tuples(st.integers(0, 3), st.integers(0, 3)), min_size=2, max_size=2))
        if data_format == "NHWC":
            explicit_paddings = [(0, 0), (padding[0][0], padding[0][1]), (padding[1][0], padding[1][1]), (0, 0)]
        else:
            explicit_paddings = [(0, 0), (0, 0), (padding[0][0], padding[0][1]), (padding[1][0], padding[1][1])]
            if data_format == "NCHW":
                explicit_paddings.insert(0, (0, 0))
                explicit_paddings.insert(1, (0, 0))

    return x, filters, strides, dilations, data_format, explicit_paddings
```
and then the decorator will look like this

```
@handle_frontend_test(
    fn_tree="tensorflow.raw_ops.Conv2D",
    x_f_d_df_ep=conv2d_inputs(),
    test_with_out=st.just(False),
)
```

Reply:
@AnnaTz  is this ok, 

```
@st.composite
def explicit_paddings(draw, x_f_d_df):
    x, filters, strides, dilations, data_format = x_f_d_df

    padding = draw(st.sampled_from(["SAME", "VALID", "EXPLICIT"]))
    if padding != "EXPLICIT":
        return None

    padding_values = draw(
        st.lists(
            st.tuples(
                st.integers(min_value=0, max_value=3),
                st.integers(min_value=0, max_value=3),
            ),
            min_size=2,
            max_size=2,
        )
    )

    if data_format == "NHWC":
        padding_values = [(0, 0), (padding_values[0][0], padding_values[0][1]), (padding_values[1][0], padding_values[1][1]), (0, 0)]
    else:
        padding_values = [(0, 0), (0, 0), (padding_values[0][0], padding_values[0][1]), (padding_values[1][0], padding_values[1][1])]

    if data_format == "NCHW":
        padding_values.insert(0, (0, 0))
        padding_values.insert(1, (0, 0))

    return padding_values

```
and then the decorator looks like this
```

@handle_frontend_test(
    fn_tree="tensorflow.raw_ops.Conv2D",
    x_f_d_df=st.builds(x_and_filters, 
                       dtypes=helpers.get_dtypes("float", full=False),
                       data_format=st.sampled_from(["NHWC", "NCHW"]),
                       padding=st.sampled_from(["SAME", "VALID", "EXPLICIT"]),
                       type="2d",
                       dilation_min=1,
                       dilation_max=1),
    explicit_paddings=explicit_paddings(),
    test_with_out=st.just(False),
)
```

Reply:
or can we do it like this, 

```
@st.composite
def explicit_paddings(draw):
    x, filters, strides, dilations, data_format, padding = _x_and_filters(draw)

    if padding == "EXPLICIT":
        pad_top = draw(st.integers(min_value=0, max_value=3))
        pad_bottom = draw(st.integers(min_value=0, max_value=3))
        pad_left = draw(st.integers(min_value=0, max_value=3))
        pad_right = draw(st.integers(min_value=0, max_value=3))

        if data_format == "NHWC":
            padding = [(0, 0), (pad_top, pad_bottom), (pad_left, pad_right), (0, 0)]
        else:
            padding = [(0, 0), (0, 0), (pad_top, pad_bottom), (pad_left, pad_right)]

        if data_format == "NCHW":
            padding.insert(0, (0, 0))
            padding.insert(1, (0, 0))

    else:
        padding = None

    return padding

```

Reply:
This last one is closer to what I'm saying, but still. Since you call `_x_and_filters` there you should return `x, filters, strides, dilations, data_format, padding` along with the padding you generate. If you call `_x_and_filters` here and then again separately in the wrapper, there is no guarantee that the `data_format` will be the same. So you should have one composite strategy that returns all the needed arguments together, called `_conv2d_helper` for example.

Reply:
so there should be a another variable as explicit_padding that returns the padding values if the `padding == "EXPLICIT"`

Reply:
yes, otherwise you still return explicit_padding  but it's an empty list

Reply:
ok sure. done

Reply:
@AnnaTz  there is some formatting problem, i'll fix that

Reply:
lint says the line is too long, and when i break the line it adds a `\` automatically, so should i use that